content/posts/api-contracts-definitons.md

```diff
@@ -0,0 +1,126 @@
+---
+title: API Contract Definitions
+author: Lena Fuhrimann
+comments: true
+date: 2022-10-03
+draft: true
+---
+
+When running a landscape of microservices, it becomes more and more important that they have reliable [service contracts](https://cloud.google.com/appengine/docs/legacy/standard/java/designing-microservice-api) between each other. An example would be a service which exposes an API to query Pokémon by their respective ID. Such contracts mostly consist of declarative interface definitions, which strongly define and type the API exposed by the respective service. As such, it is crucial that the code making up the service exactly implements the interface and therefore fulfills its side of the contract. Regressions need to be detected and changes reflected in a well communicated update to the contract. Here, we want to look at different ways of specifying contracts for what is one of the most common protocols for exposing service APIs: **HTTP**.
+
+HTTP works great as a means of communication for microservices because it is open, reliable, programming language-agnostic and works great over the wire. All of these features are crucial to modern services, as they allow engineers to change the underlying technologies (e.g., change the back end code from Python to Go) without it affecting the contract. Therefore, the API's consumers don't even need to know about the implementing technology and the providing team can take independent decisions respectively.
+
+Service contracts usually contain the following four components:
+
+- Available endpoints and operations on each endpoint
+- Operation parameters Input and output for each operation
+- Authentication methods
+- Contact information, license, terms of use and other information
+
+## Manual Specification
```

Comment:
I believe it would help the reader to group the paragraphs "Manual Specification" and "Contract First vs. Implementation First" into it's own heading like "Specification and Implementation" with a short preamble such as:

## Specification and Implementation

A service contract generally has two parts, the specification and the implementation. Ideally, these should always be in sync, as the best documentation is useless, if it does not accurately reflect the reality of the API implementation. There are generally 3 ways of keeping the two in sync:

### Manual

[...]

### Contract First

[...]

Implementation First

[...]

The manual approach is prone to drift between the implementation and specification. Using either the contract first or implementation first approach guarantee that there is a single source of truth and that the other part of the service contract is always in sync. As such, both are viable approaches. However, in general, it is preferred to write the contract first and generate implementation code from it. The reason being that when you begin implementing your service, ideally, the contract has already been defined and communicated with potential consumers of your API to allow them to work independently of your implementation. Having a human- and machine-readable contract checked into your source code repository allows you to track changes to that contract over time and additionally serves as documentation for what the implementation code does (or at least what it should do).

Reply:
Thanks. Done.